third_party/nvfuser/csrc/multidevice/aggregate_dag.h

```diff
@@ -0,0 +1,162 @@
+#pragma once
+#include <fusion.h>
+#include <ir_base_nodes.h>
+#include <ir_container.h>
+#include <ir_printer.h>
+#include <iter_visitor.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+class Group;
+class MultiGroupFusion;
+class AggregateVal;
+
+// TODO: Revisit
+using GroupPtr = std::shared_ptr<Group>;
+
+// Implements the AggregateDag. It is build from a multigroup fusion.
+// Its traversal is what defines the runtime execution.
+// The nodes of the dag are the Vals AggregateVal
+// The edges are Expr of either type 1) AggregateExpr which represent the
+// operations contained in a Group, or 2)SendRecv which represents the
+// collective operation between two processes that needs to be done at runtime.
+
+// AggregateDag is an IrContainer, built from a multigroup fusion.
+// Its traversal is what orders the runtime execution.
+class TORCH_CUDA_CU_API AggregateDag : public Fusion {
+ public:
+
+  AggregateDag(const MultiGroupFusion* MGFusion);
+
+  std::string toString();
+
+  const std::vector<Val*>& MGFusionInputs() const;
+  const std::vector<Val*>& MGFusionOutputs() const;
+
+ private:
+  const MultiGroupFusion* MGFusion_;
+
+// The following functions are a set of helper functions for build()
+  void buildAVals();
+  void buildAExpr();
+  void buildSendRecv();
+};
+
+
+// An AggregateVal is basically a Val and a Group. At runtime, since a Group is
+// typically associated with a rank, AggregateVal represents a Val which is
+// located at a rank.
+class TORCH_CUDA_CU_API AggregateVal : public Val {
+ public:
+  AggregateVal(IrBuilderPasskey passkey, Val* val, GroupPtr group);
+  AggregateVal(const AggregateVal* src, IrCloner* ir_cloner);
+
+  NVFUSER_DECLARE_CLONE
+
+  virtual std::string toString(int indent_size = 0) const override;
+  virtual std::string toInlineString(int indent_size = 0) const override;
+
+  Val* getOriginalVal() const {
+    return original_val_;
+  }
+
+  const GroupPtr getGroup() const {
+    return group_;
+  }
+
+  bool isValOutputOfGroup() const;
+
+  bool isValInputOfGroup() const;
+
+  bool sameAs(const Statement* other) const override;
+
+ private:
+  Val* original_val_;
+  GroupPtr group_;
+};
+
+// An AggregateExpr represents the Exprs that are defined inside a Group.
+class TORCH_CUDA_CU_API AggregateExpr : public Expr {
+ public:
+  using Expr::Expr;
+  AggregateExpr(IrBuilderPasskey, GroupPtr group);
+
+  NVFUSER_DECLARE_CLONE_AND_CREATE
+
+  virtual std::string toString(int indent_size = 0) const override;
+  virtual std::string toInlineString(int indent_size = 0) const override;
+
+  const char* getOpString() const override {
+    return "AggregateExpr";
+  }
+
+  bool sameAs(const Statement* other) const override;
+
+  const GroupPtr getGroup() const {
+    return group_;
+  }
+
+  void addInput(AggregateVal* input) {
+    Expr::addInput(input);
+  }
+
+  void addOutput(AggregateVal* output) {
+    Expr::addOutput(output);
+  }
```

Comment:
You are right. I had to move the calls to Expr::addInput/Output to inside AggregateExpr builder since they are private functions

Reply:
Not sure what you mean. `Expr::addInput/Output` are not private but just protected, so any subclasses can use them.

Reply:
Yes, since there are protected, the methods can be called from `AggregateExpr` (which is derived from `Expr`) but not from `AggregateDag`. This is why I had to move them (they used to be called inside `AggregateDag::.buildAExpr()`)
But I actually also find it nicer as it is now.